R/forecastHybrid_model.R

```diff
@@ -0,0 +1,86 @@
+#' models or the maximum/minimum from the candiate models. Note that `forecastHybrid` must be installed for this
+#' model wrapper to be functional.
+#' @param y Numeric vector of time points to forecast
+#' @param samples Numeric, number of samples to take.
+#' @param horizon Numeric, the time horizon over which to predict.
+#' @param model_params List of parameters to pass to `forecastHybrid::hybridModel`.
+#' @param forecast_params List of parameters to pass to `forecastHybrid:::forecast.hybridModel`.
+#' @return A dataframe of predictions (with columns representing the time horizon and rows representing samples).
+#' @export
+#' @importFrom purrr map2
+#' @importFrom dplyr bind_cols
+#' @examples \dontrun{
+#'
+#' library(forecastHybrid)
+#'
+#' ## Used on its own
+#' forecastHybrid_model(y = EpiSoon::example_obs_rts$rt,
+#'                      samples = 10, horizon = 7)
+#'
+#'
+#'## Used with non-default arguments
+#'## Note that with the current sampling from maximal confidence intervals model
+#'## Weighting using cross-validation will only have an impact whhen 1 sample is used.
+#'forecastHybrid_model(y = EpiSoon::example_obs_rts$rt,
+#'                     samples = 1, horizon = 7,
+#'                     model_params = list(cvHorizon = 7, windowSize = 7,
+#'                                   rolling = TRUE, models = "zta"))
+#'
+#'
+#' ## Used for forecasting
+#'  forecast_rt(EpiSoon::example_obs_rts,
+#'                     model = EpiSoon::forecastHybrid_model,
+#'                     horizon = 7, samples = 1)
+#'
+#'## Used for forcasting with non-default arguments
+#'forecast_rt(EpiSoon::example_obs_rts,
+#'            model = function(...){EpiSoon::forecastHybrid_model(
+#'            model_params = list(models = "zte"),
+#'            forecast_params = list(PI.combination = "mean"), ...)
+#'            },
+#'            horizon = 7, samples = 10)
+#'}
+forecastHybrid_model <- function(y = NULL, samples = NULL,
+                                 horizon = NULL, model_params = NULL,
+                                 forecast_params = NULL) {
+
+
+  check_suggests("forecastHybrid")
+
+
+  ## Fit the model
+  fitted_model <- suppressMessages(
+    suppressWarnings(
+      do.call(forecastHybrid::hybridModel, c(list(y = y), model_params))
+      )
+    )
+
+  ## Predict using the model
+  prediction <- do.call(forecastHybrid:::forecast.hybridModel,
+                        c(list(object = fitted_model, h = horizon),
+                          forecast_params))
+
+  ## Extract samples and tidy format
+  sample_from_model <- prediction
+
+  if (samples == 1) {
+    sample_from_model <- data.frame(t(as.data.frame(sample_from_model$mean)))
+    rownames(sample_from_model) <- NULL
+  }else{
+    upper <- prediction$upper[, ncol(prediction$upper)]
+    lower <-  prediction$lower[, ncol(prediction$lower)]
+    sample_from_model <- purrr::map2(lower, upper,
+                           ~ rnorm(samples, .x + (.y - .x) / 2,  (.y - .x) / 3.92))
```

Comment:
I would suggest to replace `.x + (.y - .x) / 2` by the mean. This assumes that mean and median are the same? Or is that intentional?

Reply:
`(.y - .x) / 3.92` should be replaced by `(.y - .x) / 3.289707` as we are dealing with 95% quantiles, not 97.5%. Alternatively use `mean - upper / 1.644854`

Reply:
i dont think that gives the median. The reason to use this rather than their supplied mean is that it can be weighted by the user and so not be the mean the confidence intervals refer to.

Reply:
are these quantiles or CIs or quantiles being used as CIs? 

I think they mean 95% range so 2.5 ~ 97.5 was that not your interpretation?

Reply:
retracted. I think you're right

Reply:
Regarding the mean / median: As I understood it, the mean is affected by the weighted ensemble, just the CI is computed as unweighted average / min/max. Wouldn't it therefore make sense to use the mean as is and only use the CI for the sd for now?

Reply:
I think this is correct but they don't make it very clear. As this is frequentist quantiles would be a weird thing to do considering they can't even general ensemble uncertainty.

Reply:
How can the user supply their own CI?

Reply:
because they have the standard deviation and just don't surface it. Ithink

Reply:
because they have the SD and don't surface it.

Reply:
decision?

Reply:
happy to merge as is. will make a journey down the rabbit hole to see what's there now :-)

Reply:
amazing. there is definitely a better way to do this and having a cross-validated ensemble would be nice. If we are just going to use your CRPS method less of an issue

Reply:
![image](https://user-images.githubusercontent.com/37978797/82066463-bd818480-96cf-11ea-917c-ec85cae7b206.png)
they do differ a bit. Could be relevant for Rts. maybe make an option to switch between the two? Don't feel like I can make a justified decision at this moment. If you do, I can live with either of them

Reply:
So the reason we aren't doing that is if the user changes and option those change from being unweghted by the model 
to weighted. The cis won't change though and then everything will be wrong.

Reply:
not necessarily. The variance might be little bit to small or too large, but I would argue that is still ok if you can get the mean more precise. Currently both mean and CI --> variance are obtained from an unweighted ensemble. If you change the settings, then the mean will be from a weighted ensemble and the CI not (just as good as before). So at least we get the mean a bit better.

Reply:
Also there is a difference even now without any changed settings and I don't know which one is correct

Reply:
Just wanted to flag this, so feel free to ignore :)

Reply:
happy for it to be changed to the mean. Might be best to open another pull request

Reply:
ok will do